tests/unit_tests/test_matchmaker_queue_algorithm.py

```diff
@@ -439,3 +445,129 @@ def test_make_matches_communicates_failed_attempts(p):
     assert not matches
     assert s1.failed_matching_attempts == 1
     assert s2.failed_matching_attempts == 1
+
+
+@pytest.mark.parametrize("make_teams_func", (
+    algorithm.make_teams,
+    algorithm.make_teams_from_single
+))
+@given(
+    searches=st_searches_list(max_players=1),
+    size=st.integers(min_value=1, max_value=10)
+)
+def test_make_teams_single_correct_size(searches, size, make_teams_func):
+    matched, non_matched = make_teams_func(searches, size)
+
+    for search in matched:
+        assert len(search.players) == size
+
+
+def test_make_teams_single_2v2_large_pool(p):
+    """
+    When we have a large number of players all with similar ratings, we want
+    teams to be formed by putting players with the same rating on the same team.
+    """
+
+    # Large enough so the test is unlikely to pass by chance
+    num = 40
+
+    searches = [Search([p(1000, 10, name=f"p{i}")]) for i in range(num)]
+    searches += [Search([p(500, 10, name=f"p{i}")]) for i in range(num)]
+    matched, non_matched = algorithm.make_teams_from_single(searches, size=2)
+
+    assert matched != []
+    assert non_matched == []
+
+    for search in matched:
+        p1, p2 = search.players
+        assert p1.ratings[RatingType.LADDER_1V1] == p2.ratings[RatingType.LADDER_1V1]
+
+
+def test_make_teams_single_2v2_small_pool(p):
+    """
+    When we have a small number of players, we want teams to be formed by
+    distributing players of equal skill to different teams so that we can
+    maximize the chances of getting a match.
+    """
+
+    # Try a bunch of times so it is unlikely to pass by chance
+    for _ in range(20):
+        searches = [Search([p(1000, 10, name=f"p{i}")]) for i in range(2)]
+        searches += [Search([p(500, 10, name=f"r{i}")]) for i in range(2)]
+        matched, non_matched = algorithm.make_teams_from_single(searches, size=2)
+
+        assert matched != []
+        assert non_matched == []
+
+        for search in matched:
+            p1, p2 = search.players
+            # Order doesn't matter
+            if p1.ratings[RatingType.LADDER_1V1] == (1000, 10):
+                assert p2.ratings[RatingType.LADDER_1V1] == (500, 10)
+            else:
+                assert p1.ratings[RatingType.LADDER_1V1] == (500, 10)
+                assert p2.ratings[RatingType.LADDER_1V1] == (1000, 10)
+
+
+def test_make_buckets_performance(bench, p):
+    NUM_SEARCHES = 1000
+    searches = [Search([p(1500, 500, ladder_games=1)]) for _ in range(NUM_SEARCHES)]
```

Comment:
With this it will just be a single bucket and the algorithm runs in a single pass, right?
I think random ratings make more sense here, too.